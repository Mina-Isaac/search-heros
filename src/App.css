* {
  box-sizing: border-box;
}

body,
html {
  margin: 0;
  padding: 0;
}
.App {
  height: 100vh;
  width: 100vw;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  background: linear-gradient(grey, white) fixed;
  background-attachment: fixed;
  background-size: cover;
}

.input,
.result {
  background-color: rgba(255, 255, 255, 0.4);
  border: 1px solid black;
  padding: 5px;
}

.input {
  border-radius: 5px;
  width: 100%;
}

.input:focus {
  box-shadow: 0 0 5px rgba(81, 203, 238, 1);
}

.result {
  margin: -1px 1% 0 1%;
}

.Wrapper {
  display: flex;
  flex-direction: column;
  width: 35%;
  overflow: auto;
}
@media (max-width: 768px) {
  .Wrapper {
    width: 85%;
  }
}

.result:nth-child(2) {
  border-radius: 5px 5px 0 0;
}

.result:last-child {
  border-radius: 0 0 5px 5px;
}

.result:nth-child(2):last-child {
  border-radius: 5px;
}

.alias {
  color: dimgray;
}

/* Loader styles */

.input-container {
  position: relative;
  margin: 3% 1%;
}
.icon-container {
  position: absolute;
  right: 10px;
  top: calc(50% - 10px);
}
.loader {
  position: relative;
  height: 20px;
  width: 20px;
  display: inline-block;
  animation: around 5.4s infinite;
}

@keyframes around {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}

.loader::after,
.loader::before {
  content: "";
  background: transparent;
  position: absolute;
  display: inline-block;
  width: 100%;
  height: 100%;
  border-width: 2px;
  border-color: #333 #333 transparent transparent;
  border-style: solid;
  border-radius: 20px;
  box-sizing: border-box;
  top: 0;
  left: 0;
  animation: around 0.7s ease-in-out infinite;
}

.loader::after {
  animation: around 0.7s ease-in-out 0.1s infinite;
  background: transparent;
}
