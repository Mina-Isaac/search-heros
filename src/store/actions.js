import * as actionTypes from "./constants";

const searchRequest = (query) => ({
  type: actionTypes.SEARCH_REQUEST,
  query,
});

const searchSuccess = (result) => ({
  type: actionTypes.SEARCH_SUCCESS,
  result,
});

const searchFailure = (error) => ({
  type: actionTypes.SEARCH_FAILURE,
  error,
});

export { searchRequest, searchSuccess, searchFailure };
