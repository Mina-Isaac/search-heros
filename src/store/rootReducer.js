import { combineReducers } from "redux";
import loading from "./reducers/loadingReducer";
import result from "./reducers/resultReducer";

export default combineReducers({
  loading,
  result,
});
