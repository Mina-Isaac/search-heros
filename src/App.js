import React, { useEffect, useMemo, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { searchRequest } from "./store/actions";
import useDebounce from "./Utilities/debounceHook";
import "./App.css";

const App = () => {
  const dispatch = useDispatch();
  const [query, setQuery] = useState("");
  const debouncedQuery = useDebounce(query, 500);
  const results = useSelector((state) => state.result);
  const loading = useSelector((state) => state.loading);

  useEffect(() => {
    dispatch(searchRequest(debouncedQuery));
  }, [debouncedQuery]);

  function handleChange(event) {
    setQuery(event.target.value);
  }

  const resultList = useMemo(
    () =>
      results.map((item, i) => {
        return (
          <div className="result" key={i}>
            {item.name + " "}
            <span className="alias">{item.alias}</span>
          </div>
        );
      }),
    [results]
  );

  return (
    <div className="App">
      <h1>Searchable heroes</h1>
      <br />
      <div className="Wrapper">
        <div className="input-container">
          <input
            placeholder="Search DC heros by name, alias or ID"
            className="input"
            type="text"
            value={query}
            onChange={handleChange}
          />
          {loading && (
            <div class="icon-container">
              <i class="loader"></i>
            </div>
          )}
        </div>
        {resultList}
      </div>
    </div>
  );
};

export default App;
