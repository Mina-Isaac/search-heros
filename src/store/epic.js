import { of, from } from "rxjs";
import * as actionTypes from "./constants";
import { catchError, map, filter, switchMap } from "rxjs/operators";
import { searchSuccess, searchFailure } from "./actions";

export const fetchSearchResults = (action$, state$, search) =>
  action$.pipe(
    filter((action) => action.type === actionTypes.SEARCH_REQUEST),
    switchMap((action) =>
      from(search(action.query)).pipe(
        map(searchSuccess),
        catchError((error) => of(searchFailure(error)))
      )
    )
  );
