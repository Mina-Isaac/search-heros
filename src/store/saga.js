import { call, put, takeEvery, takeLatest } from "redux-saga/effects";
import * as actions from "./actions";
import { search } from "../api";

function* fetchSearchResult(action) {
  try {
    const result = yield call(search, action.query);
    yield put(actions.searchSuccess(result));
  } catch (err) {
    yield put(actions.searchFailure(err));
  }
}

function* rootSaga() {
  yield takeLatest(actions.searchRequest().type, fetchSearchResult);
}
export default rootSaga;
