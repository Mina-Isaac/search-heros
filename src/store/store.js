import { applyMiddleware, createStore } from "redux";
import createSagaMiddleware from "redux-saga";
import rootReducer from "./rootReducer";
import rootSaga from "./saga";

const sagaMiddleware = createSagaMiddleware();
const enhancer = applyMiddleware(sagaMiddleware);
const store = createStore(rootReducer, enhancer);
sagaMiddleware.run(rootSaga);

export default store;
