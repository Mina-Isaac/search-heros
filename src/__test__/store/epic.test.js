import { StateObservable, ActionsObservable } from "redux-observable";
import { Subject } from "rxjs";
import * as actions from "../../store/actions";
import { fetchSearchResults } from "../../store/epic";
import { heroes as sampleResponse } from "../../api/db.json";
import { initialState } from "../components/App.test";

let state$;
let search;
describe("Search epic", () => {
  beforeEach(() => {
    state$ = new StateObservable(new Subject(), initialState);
    search = jest.fn();
  });

  it("should dispatch searchSuccess action when search api returns", (done) => {
    search.mockReturnValue(Promise.resolve(sampleResponse));
    const action$ = ActionsObservable.of(actions.searchRequest());
    const expectedOutputAction = actions.searchSuccess(sampleResponse);
    fetchSearchResults(action$, state$, search).subscribe((actionReceived) => {
      expect(actionReceived.type).toBe(expectedOutputAction.type);
      done();
    });
  });
  it("should dispatch searchFailure action when the request fails", (done) => {
    search.mockReturnValue(Promise.reject(new Error("Some error")));
    const action$ = ActionsObservable.of(actions.searchRequest());
    const expectedOutputAction = actions.searchFailure(new Error());
    fetchSearchResults(action$, state$, search).subscribe((actionReceived) => {
      expect(actionReceived.type).toBe(expectedOutputAction.type);
      done();
    });
  });
});
