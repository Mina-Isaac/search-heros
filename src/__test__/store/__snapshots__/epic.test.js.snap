// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render without issues renders correctly 1`] = `
<div
  className="App"
>
  <h1>
    Searchable heroes
  </h1>
  <br />
  <div
    className="Wrapper"
  >
    <div
      className="input-container"
    >
      <input
        className="input"
        onChange={[Function]}
        placeholder="Search DC heros by name, alias or ID"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
